### ISSUE TITLE
New Teams Mute Functionality

### ISSUE THREAD

**Issue by ryanbuening** [tdalon/ahk#18]
Hello. Creating an issue here as requested [here](https://www.autohotkey.com/boards/viewtopic.php?p=543663#p543663). I was previously able to use the script below to mute/unmute Teams v1 meetings . The script was/is convenient as no other libraries, tools, or extra scripts were necessary. With the release of the new [Teams v2](https://adoption.microsoft.com/en-us/new-microsoft-teams/), the script no longer works.

```ahk
Pause::
; Get IDs for all teams windows
WinGet, id, list, ahk_exe Teams.exe		

; Loop through IDs of all teams windows
Loop, %id% 
{
	this_ID := id%A_Index%

	; Get the title of the current window
	WinGetTitle, Title, ahk_id %this_ID%
	
	; Make sure title is not the notification
	if Title <> Microsoft Teams Notification	
	{
		; Screen sharing win uses null title, make sure the win does not have a null title
		if Title <>
		{
			; This should be the correct win, activate it
			WinActivate, ahk_id %this_ID%

			; Send Ctrl Shift M shortcut
			Send, ^M

			; There are two teams windows, the main win and the meeting win, break the loop so that the mute commmand doesnt get sent twice
			break
		}
	}
}
```

**Comment by tdalon** [comment]
I have pushed an update and tested the mute implementation (partly)

Look at the code in the repo.
Short explanation see https://tdalon.blogspot.com/2023/10/update-teams-powertools-for-teams-new.html

**Comment by ryanbuening** [comment]
@tdalon - I have implemented your changes into my script, and I appear to be having an issue in `Teams_GetMeetingWindow()`. I am getting the `Could not find Meeting Window!` alert. It seems as though the logic is unable to set the `TeamsEl` variable...

```ahk
Teams_GetMeetingWindow(Maximize:=false, Activate:=false){
UIA := UIA_Interface()
TeamsExe := Teams_GetExeName()

WinGet, Win, List, ahk_exe %TeamsExe%
Loop %Win% {
    WinId := Win%A_Index%
	MsgBox "WinId" + %WinId% ; This does get logged twice when I'm in a meeting (and once when I'm not) and appears to have a value (e.g. 0xa0672)
    TeamsEl := UIA.ElementFromHandle(WinId)
	MsgBox "TeamsEl" + %TeamsEl% ; <-- The TeamsEl variable is empty here
    If Teams_IsMeetingWindow(TeamsEl)  {
        If (Maximize) {
            El := TeamsEl.FindFirstByNameAndType("Navigate back to call window.", "button") ; 
            Tooltip("max")
            If !Activate
                WinGet, prevWinId, ID, A
            
            If El {
                El.Click()
                Sleep 500
            }

            WinGet, WinId, ID, A ; Weird Teams Client Behavior: WinId changes after maximization
                
            If !Activate
                WinActivate, ahk_id %prevWinId%
        }

        If Activate {
            ;MsgBox Activate %WinId%
	    Tooltip("Activate")
            WinActivate, ahk_id %WinId% ; sometimes does not work
        }
        return WinId
	Tooltip("???")
    }

    ;MsgBox % TeamsEl.DumpAll()
} ; End Loop

TrayTip, Could not find Meeting Window! , No active Teams meeting window found!,,0x2
} ; eofun
```

Any ideas?

**Comment by tdalon** [comment]
Hi Ryan. I have updated/refactored the code a bit. Please take the last version in the repo and retry it.

The TeamsEl is an UIA Object that you can debug view with e.g. TeamsEl.DumpAll()

I guess you are using an English client, right? If not you might need to configure a bit and I would need to explain more.

**Comment by ryanbuening** [comment]
@tdalon - appears to be working now. Thanks!

**Comment by dooniem** [comment]
Thanks for taking the time to rewrite the mute-function to also work in Teams v2.
But it seems it's only partly backwards compatible?

There are still some bugs in Teams v2, so I happened to switch back to classic.
Your `Teams_IsNew(){ ` only checks if ms-teams.exe exists in the default v2 directory or not. In my situation it does, but I'm running Classic. I had to patch a little quick fix, but maybe you will find a smoother way:

```
Teams_IsNew(){
    static IsNew
    If !(IsNew = "")
        return IsNew
    
	Process, Exist, Teams.exe
	If (!ErrorLevel= 0) {
		;Classic Teams
		return IsNew
	}
	Else{
		;Teams v2
		IsNew:= 1
	}
    return IsNew
} ;eofun
```

(My understanding was that IsNew returned either blank or 1, so I tried reproducing your approach)
